### PR TITLE
[WIP] Initial implementation of attestation verification

### PR DESCRIPTION
**Pull request by vishal-chdhry**
Initial implementation of attestation verification based on https://github.com/nirmata/kyverno-notation-aws/pull/12

cc @JimBugwadia @realshuting

### CODE REVIEW

File: verifier/verify.go [hunk 1]
```diff
@@ -187,24 +218,141 @@ func (v *verifier) verifyImages(ctx context.Context, images *ImageInfos) ([]byte
...
+		imageAttestations, err := v.verifyAttestation(ctx, image, list)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to verify attestations")
+		}
+		attestations = append(attestations, imageAttestations...)
+	}
+	return attestations, nil
+}
+
+func (v *verifier) verifyAttestation(ctx context.Context, image string, attestationList map[string]bool) ([]Attestation, error) {
```
[the remark this answers is not in the document]
Comment (Vishal-Chdhry) [reply verifier/verify.go:233]: What it does is
1. Fetches all the referrers for the `image`
2. Loops over the list of referrers found
3. If the referrer is present in the `attestationList`, it calls `verifyReference` [here](https://github.com/nirmata/kyverno-notation-verifier/pull/1/files/e5807c238daaa0ebb31580e9fbdaff7e261e35eb#diff-0e1aaf1238ba75fa4dcdd1e89cb1cc19a990a1a604ed505cd0d2982ab062c66dR271) to do a verification of the notary signature

File: verifier/response.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+func (r *Response) GetResponse() ResponseData {
+	return r.ResponseData
+}
+
+func (r *Response) GetImageList() map[string]AttestationList {
+	return r.ImageList
+}
+
+func (r *Response) VerificationFailed(msg string) ([]byte, error) {
```
Comment (JimBugwadia) [inline verifier/response.go:39]: does this need to be public?

File: verifier/response.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+	r.ResponseData.Attestations = nil
+
+	data, err := json.MarshalIndent(r.ResponseData, "  ", "  ")
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to marshal response")
+	}
+	return data, nil
+}
+
+func (r *Response) AddImage(img *ImageInfo) {
```
Comment (JimBugwadia) [inline verifier/response.go:52]: Same question - on scope....

File: verifier/verify.go [hunk 4]
```diff
@@ -125,86 +129,179 @@ func (v *verifier) Stop() {
...
+		attestations[i] = Attestation{
+			Type:    referrer.ArtifactType,
+			Image:   referrerRef,
+			Payload: payload,
+		}
+	}
+	return attestations, nil
+}
+
+func (v *verifier) extractPayload(ctx context.Context, repoRef name.Reference, desc v1.Descriptor, options ...gcrremote.Option) (map[string]interface{}, error) {
```
Comment (JimBugwadia) [inline verifier/verify.go:244]: If this method is also doing a fetch, we can perhaps rename it to indicate that? e.g. `fetchPayload` or `fetchAndExtractPayload`?

**Review by JimBugwadia — COMMENTED** [review]
minor comments....

For all methods that are not exposed to the external clients, please make the them private and consider moving to an `internal` package.

**Comment by JimBugwadia** [comment]
@vishal-chdhry - can we please follow the Golang project org guidelines and move all code to `pkg\notation` or similar?

Also, please hide all implementation methods and separate out the minimal interface we want to expose to clients.

**Comment by vishal-chdhry** [comment]
@JimBugwadia Sure, I will update the structure of the project and the active PRs using this guide: https://github.com/golang-standards/project-layout and I will reach back as soon as I am done

**Comment by vishal-chdhry** [comment]
replaced by #6

**Pull request closed without merging**